File: 3. Data Encryption Standard (DES)/DES.go
package main

import (
	"bytes"
	"crypto/cipher"
	"crypto/des"
	"fmt"
)

func main() {
	originalText := "this is secret text$$$"
	fmt.Println(originalText)
	mytext := []byte(originalText)

	key := []byte{0xBC, 0xBC, 0xBC, 0xBC, 0xBC, 0xBC, 0xBC, 0xBC}
	iv := []byte{0xBC, 0xBC, 0xBC, 0xBC, 0xBC, 0xBC, 0xBC, 0xBC}

	cryptoText, _ := encrypt(key, iv, mytext)
	fmt.Println(string(cryptoText))
	decryptedText, _ := decrypt(key, iv, cryptoText)
	fmt.Println(string(decryptedText))

}

func encrypt(key, iv, mytext []byte) ([]byte, error) {

	block, err := des.NewCipher([]byte(key))
	if err != nil {
		panic(err)
	}
	fmt.Println(block, block.BlockSize())
	blockSize := block.BlockSize()
	data := add_padding(mytext, blockSize)
	fmt.Println(data)
	mode := cipher.NewCBCEncrypter(block, iv)
	crypted := make([]byte, len(data))
	mode.CryptBlocks(crypted, []byte(data))
	return crypted, nil

}

func decrypt(key, iv, cryptoText []byte) ([]byte, error) {
	block, err := des.NewCipher(key)

	if err != nil {
		return nil, err
	}

	blockMode := cipher.NewCBCDecrypter(block, iv)
	origData := make([]byte, len(cryptoText))
	blockMode.CryptBlocks(origData, cryptoText)
	origData = remove_padding(origData)
	return origData, nil
}

func add_padding(src []byte, blockSize int) []byte {
	padding := blockSize - len(src)%blockSize
	padtext := bytes.Repeat([]byte{byte(padding)}, padding)
	return append(src, padtext...)
}

func remove_padding(src []byte) []byte {
	length := len(src)
	unpadding := int(src[length-1])
	return src[:(length - unpadding)]
}

File: 3. Data Encryption Standard (DES)/DES.py
from Crypto.Cipher import DES

def add_padding(text,blocksize=64):
    pad_len = blocksize - (len(text) % blocksize)
    padding = '$' * pad_len
    return text + padding

def remove_padding(text):
    counter = 0

    print(text[::-1])
    for c in text[::-1]:
        if c == "$":
            counter+=1
        else:
            break
    print("t",text)
    return text[:-counter]

def encrypt(plain_text,key):
    des = DES.new(key,DES.MODE_ECB)
    return des.encrypt(plain_text)

def decrypt(cipher_text,key):
    des = DES.new(key,DES.MODE_ECB)
    return des.decrypt(cipher_text).decode('UTF-8')


if __name__ == '__main__':
    key = "secretaa"
    plain_text = 'This is the secret message we want to encrypt'
    plain_text = add_padding(plain_text)
    cipher_text = encrypt(plain_text,key)
    print(cipher_text)
    decrypted_message = decrypt(cipher_text,key)
    print("d",decrypted_message)
    decrypted_message = remove_padding(decrypted_message)
    print(decrypted_message)
    